src/interchange/src/avro/schema.rs

```diff
@@ -60,71 +91,85 @@ fn validate_schema_1(schema: SchemaNode) -> anyhow::Result<Vec<(ColumnName, Colu
-                                .1
-                                .map(|full_name| full_name.base_name().to_owned())
-                        })
-                        .unwrap_or_else(|| "?column?".into())
-                } else {
-                    // There are multiple non-null variants in the
-                    // union, so we need to invent field names for
-                    // each variant.
-                    base_name
-                        .map(|n| format!("{}{}", n, i + 1))
-                        .or_else(|| {
-                            v.get_piece_and_name(schema.root)
-                                .1
-                                .map(|full_name| full_name.base_name().to_owned())
-                        })
-                        .unwrap_or_else(|| "?column?".into())
-                };
+            let name = if vs.len() == 1 || (vs.len() == 2 && vs.iter().any(is_null)) {
+                // There is only one non-null variant in the
+                // union, so we can use the field name directly.
+                base_name
+                    .map(|n| n.to_owned())
+                    .or_else(|| {
+                        v.get_piece_and_name(schema.root)
+                            .1
+                            .map(|full_name| full_name.base_name().to_owned())
+                    })
+                    .unwrap_or_else(|| "?column?".into())
+            } else {
+                // There are multiple non-null variants in the
+                // union, so we need to invent field names for
+                // each variant.
+                base_name
+                    .map(|n| format!("{}{}", n, i + 1))
+                    .or_else(|| {
+                        v.get_piece_and_name(schema.root)
+                            .1
+                            .map(|full_name| full_name.base_name().to_owned())
+                    })
+                    .unwrap_or_else(|| "?column?".into())
+            };
 
-                // If there is more than one variant in the union,
-                // the column's output type is nullable, as this
-                // column will be null whenever it is uninhabited.
-                let ty = validate_schema_2(seen_avro_nodes, node)?;
-                columns.push((name.into(), ty.nullable(vs.len() > 1)));
-                if let Some(named_idx) = named_idx {
-                    seen_avro_nodes.remove(&named_idx);
-                }
+            // If there is more than one variant in the union,
+            // the column's output type is nullable, as this
+            // column will be null whenever it is uninhabited.
+            let ty = validate_schema_2(seen_avro_nodes, node)?;
+            columns.push((name.into(), ty.nullable(vs.len() > 1)));
+            if let Some(named_idx) = named_idx {
+                seen_avro_nodes.remove(&named_idx);
             }
         }
-        Ok(columns)
+    }
+    Ok(columns)
+}
```

Comment:
did this function change at all other than refactoring?

Reply:
The diff is way cleaner if you view with whitespace suppressed! https://github.com/MaterializeInc/materialize/pull/13255/files?w=1